(kubectl debug): Support debugging via files

ardaguclu opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Currently `kubectl debug` only supports passing names in command line.
However, users might want to pass resources in files by passing `-f` flag like
in all other kubectl commands.

This PR adds this ability.

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Add new -f flag into debug command to be used passing pod or node files instead explicit names.
```

ardaguclu commented:
/assign @verb 

Could you please take a look at this change?. What is your opinions about it?. Thanks.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -325,65 +333,67 @@ func (o *DebugOptions) Run(f cmdutil.Factory, cmd *cobra.Command) error {
-					Quiet:     o.Quiet,
-				},
-				CommandName: cmd.Parent().CommandPath() + " attach",
+	var (
```
all the rest is just indentation change

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -691,18 +701,15 @@ func (o *DebugOptions) computeDebugContainerName(pod *corev1.Pod) string {
 	if len(o.Container) > 0 {
 		return o.Container
 	}
-	name := o.Container
```
It is precise that name is empty.

ardaguclu commented:
/retest

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -78,6 +78,10 @@ var (
 		# (requires the EphemeralContainers feature to be enabled in the cluster)
 		kubectl debug mypod -it --image=busybox
 
+		# Create an interactive debugging session in pod in pod.yamlfile and immediately attach to it.
```
```suggestion
		# Create an interactive debugging session for the pod in the file pod.yaml and immediately attach to it.
```

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -163,6 +170,7 @@ func NewCmdDebug(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.
 	}
 
 	addDebugFlags(cmd, o)
+	cmdutil.AddFilenameOptionFlags(cmd, &o.FilenameOptions, "identifying the resource to debug")
```
`kubectl exec` only supports `--filename`. Do you think there's a reason we should support kustomize, recursive, etc with kubectl debug?

ardaguclu replied:
Please correct me if I'm wrong,  but I think, `exec` can also support kustomize, recurive, etc.?.

ardaguclu replied:
I asked in sig-cli https://kubernetes.slack.com/archives/C2GL57FJ4/p1658931680284329

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -325,65 +333,67 @@ func (o *DebugOptions) Run(f cmdutil.Factory, cmd *cobra.Command) error {
-	if err := r.Err(); err != nil {
-		return err
+		SingleResourceType().
+		Flatten()
```
What's the reasoning behind SingleResourceType and Flatten?

ardaguclu replied:
Just to guarantee that we are debugging only one resource at a time as we're discussing in here https://github.com/kubernetes/kubernetes/pull/111453#discussion_r931103797

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -325,65 +333,67 @@ func (o *DebugOptions) Run(f cmdutil.Factory, cmd *cobra.Command) error {
+		SingleResourceType().
+		Flatten()
+	if len(o.FilenameOptions.Filenames) > 0 {
+		builder.FilenameParam(o.explicitNamespace, &o.FilenameOptions)
```
It looks like it's fine to add `FilenameParam` unconditionally, as `kubectl exec` does.

ardaguclu replied:
Good point, let me try that and assure that it is working in that way

ardaguclu replied:
I asked in sig-cli https://kubernetes.slack.com/archives/C2GL57FJ4/p1658931680284329

verb commented on staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:
```diff
@@ -325,65 +333,67 @@ func (o *DebugOptions) Run(f cmdutil.Factory, cmd *cobra.Command) error {
-		}
+	if len(infos) != 1 {
+		klog.V(4).Infof("Objects: %#v", infos)
+		return fmt.Errorf("you can debug only one resource at a time")
```
why disallow debugging multiple objects?

ardaguclu replied:
Probably I made a mistake by assuming that `kubectl debug` supports only one resource at a time. All other changes I made is to guarantee that. Does that mean `kubectl debug` support multiple debugging?. What happens if user pass `-it` flag in multiple resources?

ardaguclu replied:
After thinking a little bit; there is no way passing multiple resources in `kubectl debug` by name. Up to now, we always support one resource already. But with `-f` flag, users can pass multiple resources and I did these changes to prevent that.

verb replied:
iirc, we always supported multiple targets in `kubectl debug`. In the case of `-it` kubectl would proceed in order, connecting to the terminal and proceeding to the next target when the connection returned.

ardaguclu replied:
Sounds good, thanks for clarification :)

verb reviewed (commented):
Thanks for the PR! In general I'm in favor of adding a filename flag because `kubectl exec` supports it.

This PR seems to include other `kubectl debug` behavior changes. Are these necessary to support filenames or are they unrelated?

ardaguclu commented:
@verb could you PTAL one more round?, I did suggested changes. Thanks.

ardaguclu commented:
@verb would you mind have a look at this change?.

ardaguclu commented:
/triage accepted

ardaguclu commented:
@soltysh @KnVerey could you please take a look at this change?. Thanks.

oomichi commented:
The original `kubectl debug` expects a pod name is passed in the command line.
If the specified file contains multiple resources, how to handle the case?

/cc @oomichi

verb commented:
Sorry for the delay here!

/lgtm

ardaguclu commented:
/priority backlog

ardaguclu commented:
> The original `kubectl debug` expects a pod name is passed in the command line. If the specified file contains multiple resources, how to handle the case?
> 
> /cc @oomichi

@oomichi My first attempt was disallowing multiple objects but we had a conversation with @verb in here https://github.com/kubernetes/kubernetes/pull/111453#discussion_r931081118. Although it might not make sense when passing `-it` flag, it might be useful for ephemeral containers for multiple resources.

ardaguclu committed:
(kubectl debug): Support debugging via files

Currently `kubectl debug` only supports passing names in command line.
However, users might want to pass resources in files by passing `-f` flag like
in all other kubectl commands.

This PR adds this ability.

ardaguclu commented:
@verb after the new profiles PR was merged, there was a merge conflict and I rebased the PR. Could you PTAL one more time?. Thanks.

ardaguclu committed:
Add explicitNamespace into ignore list in debug unit tests

Since, `explicitNamespace` is set by kubeconfig automatically, we can
safely add this field into ignore list.

ardaguclu commented:
Unit test failure is not related to this change;

/test pull-kubernetes-unit

verb reviewed (approved):
/lgtm
